docs/notebooks/multi_trust_region.pct.py

```diff
@@ -0,0 +1,164 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: -all
+#     custom_cell_magics: kql
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.2
+#   kernelspec:
+#     display_name: .venv_310
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Batch trust region Bayesian optimization
+
+# %%
+import numpy as np
+import tensorflow as tf
+
+np.random.seed(1793)
+tf.random.set_seed(1793)
+
+# %% [markdown]
+# ## Define the problem and model
+#
+# You can use trust regions for Bayesian optimization in much the same way as we used EGO and EI in the [introduction notebook](expected_improvement.ipynb). Since the setup is much the same as in that tutorial, we'll skip over most of the detail.
+
+# %%
+import trieste
+from trieste.objectives import Branin
+
+branin = Branin.objective
+search_space = Branin.search_space
+
+num_initial_data_points = 10
+initial_query_points = search_space.sample(num_initial_data_points)
+observer = trieste.objectives.utils.mk_observer(branin)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# As usual, we'll use Gaussian process regression to model the function. Note that we set the likelihood variance to a small number because we are dealing with a noise-free problem.
+
+# %%
+from trieste.models.gpflow import GaussianProcessRegression, build_gpr
+
+gpflow_model = build_gpr(initial_data, search_space, likelihood_variance=1e-7)
+model = GaussianProcessRegression(gpflow_model)
+
+
+# %% [markdown]
+# ## Create the batch trust region acquisition rule
+#
+# We achieve Bayesian optimization with trust region by specifying `MultiTrustRegionBox` as the acquisition rule.
+#
+# This rule requires an initial number `num_query_points` of sub-spaces (or trust regions) to be provided and performs optimization in parallel across all these sub-spaces. Each region contributes one query point, resulting in each acquisition step collecting `num_query_points` points overall. As the optimization process continues, the bounds of these sub-spaces are dynamically updated.
+#
+# In addition, this rule requires the specification of a batch aquisition base-rule for performing optimization; for our example we use `EfficientGlobalOptimization` coupled with `ParallelContinuousThompsonSampling`.
+#
+# Note: the number of sub-spaces/regions must match the number of batch query points.
+
+# %%
+num_query_points = 5
+
+init_subspaces = [
+    trieste.acquisition.rule.TrustRegionBox(search_space)
+    for _ in range(num_query_points)
+]
+base_rule = trieste.acquisition.rule.EfficientGlobalOptimization(  # type: ignore[var-annotated]
+    builder=trieste.acquisition.ParallelContinuousThompsonSampling(),
+    num_query_points=num_query_points,
+)
+acq_rule = trieste.acquisition.rule.MultiTrustRegionBox(
+    init_subspaces, base_rule
+)
+
+# %% [markdown]
+# ## Run the optimization loop
```

Comment:
I think I would retain the boxes to illustrate what is going on with trust regions, its a nice visualisation and its useful to have that piece of code when developing new variants - its also a cool visualisation, we should have some notebooks with those as well :)

@vpicheny what do you think?

Reply:
we can add some new code to support that in experimental plotting package...

Reply:
I think it would make sense,  with add_patch(matplotlib.patches.Rectangle(...)) this is very simple and compact.

A notebook without visualisation of the TRs would not add as much value IMO

Reply:
I have added the visualisation at the end. It is not particularly compact though, since it is plotting the contours/points (re-using existing functions) and creating the gif. Would it be useful to move any of it to the plotting package?